Instrument informers

dgrisonnet opened the pull request:
##### Instrument informers' lister watcher

To be able to instrument the informers' lister watcher, I had to move away from the previous informer factory implementation and go back to vanilla informer because they provide a way to hook into the lister watcher.

For the instrumentation I used kube-state-metrics watch package and more precisely the [ListWatchMetrics](https://pkg.go.dev/k8s.io/kube-state-metrics@v1.9.7/pkg/watch?tab=doc#ListWatchMetrics) structure and introduced the 2 following metrics with the result label being either `success` or `error`:
```
kube_events_exporter_watch_total{result="", resource="Event"}
kube_events_exporter_list_total{result="", resource="Event"}
```


Along with that, I exported the instrumented event informer implementation in `pkg/informer`

##### Instrument event handlers

To instrument event handlers, I had to write my own custom handler that I also exported in the `pkg/informer` package.
That being said, I introduced the following metric, with the event label being either `add` or `update`.
```
kube_events_exporter_handled_resources_total{event="", resource="Event"}
```

##### Cardinality

This changes is introducing 6 new series.

Fixes #7

lilic commented on pkg/informer/event.go:
```diff
@@ -0,0 +1,64 @@
+	"k8s.io/client-go/informers/internalinterfaces"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	ksmwatch "k8s.io/kube-state-metrics/pkg/watch"
```
Any reason why you choose to import this vs just recreate these functions, you need to rewrite them anyways to match the options?

dgrisonnet replied:
I chose to import it because I wanted to use existing code to instrument the informers and the ksm watch package was fairly easy to integrate.
But you are very right that it is not abstracting much, I only gain the instrumented List/Watch functions from it.

Considering that, it might not be worthwhile to import ksm and I should just rewrite the package here.

lilic commented on pkg/informer/event-handler_test.go:
```diff
@@ -0,0 +1,154 @@
+/*
```
The file name should be event_handler_test.go?

dgrisonnet replied:
Indeed, I kept some habits from c++

lilic commented on pkg/informer/event-handler.go:
```diff
@@ -0,0 +1,59 @@
+/*
```
This should be event_handler.go file name

lilic commented on pkg/informer/event-handler_test.go:
```diff
@@ -0,0 +1,154 @@
+	"k8s.io/client-go/tools/cache"
+)
+
+func TestInstrumentedEventHandler(t *testing.T) {
```
Nice amount of tests! 👏

dgrisonnet commented:
@lilic I addressed your comments and removed kube-state-metrics dependency. In order to do that, I added `pkg/informer/listwatch.go` and `pkg/informer/listwatch_test.go`. Please have a look at these files.

lilic commented on pkg/informer/event_handler.go:
```diff
@@ -0,0 +1,75 @@
+	metrics := &EventHandlerMetrics{
+		HandledTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "kube_events_exporter_handled_resources_total",
```
kube_events_exporter_handled_resources_total why do you think this metric is useful?

dgrisonnet replied:
I thought it might be insightful, but giving it another thought, I don't think this metric is useful.

lilic commented on pkg/informer/listwatch.go:
```diff
@@ -0,0 +1,104 @@
+	metrics := &ListWatchMetrics{
+		ListTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "kube_events_exporter_list_total",
```
I would like you to have a look at the following two patterns:
- kube_events_exporter_list_total of result and resource in the labels
vs 
- kube_events_exporter_list_total kube_events_exporter_errors_list_total of the result being a seperate metric and resource still being in the label.

There are some philosophies on this that would be good for you to read. Feel free to ask me or @simonpasquier if you can't find anything on this, he has made the contribution to prom-operator where he is using the later one.

dgrisonnet replied:
Indeed, I found some resources in that regard and thus decided to go with:
- kube_events_exporter_list_total
- kube_events_exporter_list_failed_total
- kube_events_exporter_watch_total
- kube_events_exporter_watch_failed_total

While I was at it, I also removed the resource label as I did not find it very useful in my particular case.

lilic replied:
Sounds good, is it ready for another review?

dgrisonnet replied:
Yes it is please have a new look at it

dgrisonnet commented on internal/collector/event.go:
```diff
@@ -22,93 +22,84 @@ import (
 
-	collector.initInformers()
-
+	for iNs := range opts.InvolvedObjectNamespaces {
```
Indeed, I will put the logic back in initInformers
